File: app/src/main/java/com/simplecity/amp_library/ui/fragments/SuggestedFragment.java
package com.simplecity.amp_library.ui.fragments;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.annimon.stream.Optional;
import com.annimon.stream.Stream;
import com.bumptech.glide.RequestManager;
import com.simplecity.amp_library.R;
import com.simplecity.amp_library.ShuttleApplication;
import com.simplecity.amp_library.dagger.module.FragmentModule;
import com.simplecity.amp_library.model.Album;
import com.simplecity.amp_library.model.AlbumArtist;
import com.simplecity.amp_library.model.Playlist;
import com.simplecity.amp_library.model.Song;
import com.simplecity.amp_library.model.SuggestedHeader;
import com.simplecity.amp_library.ui.adapters.LoggingViewModelAdapter;
import com.simplecity.amp_library.ui.adapters.ViewType;
import com.simplecity.amp_library.ui.detail.PlaylistDetailFragment;
import com.simplecity.amp_library.ui.dialog.UpgradeDialog;
import com.simplecity.amp_library.ui.modelviews.AlbumView;
import com.simplecity.amp_library.ui.modelviews.EmptyView;
import com.simplecity.amp_library.ui.modelviews.HorizontalRecyclerView;
import com.simplecity.amp_library.ui.modelviews.SuggestedHeaderView;
import com.simplecity.amp_library.ui.modelviews.SuggestedSongView;
import com.simplecity.amp_library.ui.views.SuggestedDividerDecoration;
import com.simplecity.amp_library.utils.ComparisonUtils;
import com.simplecity.amp_library.utils.DataManager;
import com.simplecity.amp_library.utils.LogUtils;
import com.simplecity.amp_library.utils.MenuUtils;
import com.simplecity.amp_library.utils.MusicUtils;
import com.simplecity.amp_library.utils.Operators;
import com.simplecity.amp_library.utils.PermissionUtils;
import com.simplecity.amp_library.utils.ShuttleUtils;
import com.simplecityapps.recycler_adapter.adapter.ViewModelAdapter;
import com.simplecityapps.recycler_adapter.model.ViewModel;
import com.simplecityapps.recycler_adapter.recyclerview.RecyclerListener;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class SuggestedFragment extends BaseFragment implements
        SuggestedHeaderView.ClickListener,
        AlbumView.ClickListener {

    public interface SuggestedClickListener {

        void onAlbumArtistClicked(AlbumArtist albumArtist, View transitionView);

        void onAlbumClicked(Album album, View transitionView);
    }

    public class SongClickListener implements SuggestedSongView.ClickListener {

        List<Song> songs;

        public SongClickListener(List<Song> songs) {
            this.songs = songs;
        }

        @Override
        public void onSongClick(Song song, SuggestedSongView.ViewHolder holder) {
            MusicUtils.playAll(songs, songs.indexOf(song), true, (String message) -> Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show());
        }

        @Override
        public boolean onSongLongClick(Song song) {
            return false;
        }

        @Override
        public void onSongOverflowClicked(View v, Song song) {
            PopupMenu popupMenu = new PopupMenu(getContext(), v);
            MenuUtils.setupSongMenu(popupMenu, false);
            popupMenu.setOnMenuItemClickListener(MenuUtils.getSongMenuClickListener(
                    getContext(),
                    song,
                    taggerDialog -> {
                        if (!ShuttleUtils.isUpgraded()) {
                            UpgradeDialog.getUpgradeDialog(getActivity()).show();
                        } else {
                            taggerDialog.show(getFragmentManager());
                        }
                    },
                    null,
                    null));
            popupMenu.show();
        }
    }

    private static final String TAG = "SuggestedFragment";

    private static final String ARG_PAGE_TITLE = "page_title";

    private RecyclerView recyclerView;

    private ViewModelAdapter adapter;

    private CompositeDisposable refreshDisposables = new CompositeDisposable();

    @Nullable
    private Disposable setItemsDisposable;

    @Inject
    RequestManager requestManager;

    private HorizontalRecyclerView favoriteRecyclerView;
    private HorizontalRecyclerView mostPlayedRecyclerView;

    @Nullable
    private SuggestedClickListener suggestedClickListener;

    public SuggestedFragment() {
    }

    public static SuggestedFragment newInstance(String pageTitle) {
        SuggestedFragment fragment = new SuggestedFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PAGE_TITLE, pageTitle);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (getParentFragment() instanceof SuggestedClickListener) {
            suggestedClickListener = (SuggestedClickListener) getParentFragment();
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();

        suggestedClickListener = null;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ShuttleApplication.getInstance().getAppComponent()
                .plus(new FragmentModule(this))
                .inject(this);

        adapter = new LoggingViewModelAdapter("SuggestedFragment");
        mostPlayedRecyclerView = new HorizontalRecyclerView("SuggestedFragment - mostPlayed");
        favoriteRecyclerView = new HorizontalRecyclerView("SuggestedFragment - favorite");
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        if (recyclerView == null) {
            recyclerView = (RecyclerView) inflater.inflate(R.layout.fragment_suggested, container, false);
            recyclerView.addItemDecoration(new SuggestedDividerDecoration(getResources()));
            recyclerView.setRecyclerListener(new RecyclerListener());

            int spanCount = ShuttleUtils.isTablet() ? 12 : 6;

            GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), spanCount);
            gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    if (!adapter.items.isEmpty() && position >= 0) {
                        ViewModel item = adapter.items.get(position);
                        if (item instanceof HorizontalRecyclerView
                                || item instanceof SuggestedHeaderView
                                || (item instanceof AlbumView && item.getViewType() == ViewType.ALBUM_LIST)
                                || (item instanceof AlbumView && item.getViewType() == ViewType.ALBUM_LIST_SMALL)
                                || item instanceof EmptyView) {
                            return spanCount;
                        }
                        if (item instanceof AlbumView && item.getViewType() == ViewType.ALBUM_CARD_LARGE) {
                            return 3;
                        }
                    }

                    return 2;
                }
            });

            recyclerView.setLayoutManager(gridLayoutManager);
        }
        if (recyclerView.getAdapter() != adapter) {
            recyclerView.setAdapter(adapter);
        }

        return recyclerView;
    }

    @Override
    public void onResume() {
        super.onResume();

        refreshAdapterItems();
    }

    Observable<List<ViewModel>> getMostPlayedViewModels() {
        return Playlist.mostPlayedPlaylist
                .getSongsObservable()
                .map(songs -> {
                    if (!songs.isEmpty()) {
                        List<ViewModel> viewModels = new ArrayList<>();

                        SuggestedHeader mostPlayedHeader = new SuggestedHeader(getString(R.string.mostplayed), getString(R.string.suggested_most_played_songs_subtitle), Playlist.mostPlayedPlaylist);
                        SuggestedHeaderView mostPlayedHeaderView = new SuggestedHeaderView(mostPlayedHeader);
                        mostPlayedHeaderView.setClickListener(this);
                        viewModels.add(mostPlayedHeaderView);

                        viewModels.add(mostPlayedRecyclerView);

                        Collections.sort(songs, (a, b) -> ComparisonUtils.compareInt(b.playCount, a.playCount));
                        SongClickListener songClickListener = new SongClickListener(songs);

                        mostPlayedRecyclerView.viewModelAdapter.setItems(Stream.of(songs)
                                .map(song -> {
                                    SuggestedSongView suggestedSongView = new SuggestedSongView(song, requestManager);
                                    suggestedSongView.setClickListener(songClickListener);
                                    return (ViewModel) suggestedSongView;
                                })
                                .limit(20)
                                .toList());

                        return viewModels;
                    } else {
                        return Collections.emptyList();
                    }
                });
    }

    Observable<List<ViewModel>> getRecentlyPlayedViewModels() {
        return Playlist.recentlyPlayedPlaylist
                .getSongsObservable()
                .flatMap(songs -> Observable.just(Operators.songsToAlbums(songs)))
                .flatMapSingle(albums -> Observable.fromIterable(albums)
                        .sorted((a, b) -> ComparisonUtils.compareLong(b.lastPlayed, a.lastPlayed))
                        .flatMapSingle(album ->
                                // We need to populate the song count
                                album.getSongsSingle()
                                        .map(songs -> {
                                            album.numSongs = songs.size();
                                            return album;
                                        })
                                        .filter(a -> a.numSongs > 0)
                                        .toSingle()
                        )
                        .sorted((a, b) -> ComparisonUtils.compareLong(b.lastPlayed, a.lastPlayed))
                        .take(6)
                        .toList()
                )
                .map(albums -> {
                    if (!albums.isEmpty()) {
                        List<ViewModel> viewModels = new ArrayList<>();

                        SuggestedHeader recentlyPlayedHeader = new SuggestedHeader(getString(R.string.suggested_recent_title), getString(R.string.suggested_recent_subtitle), Playlist.recentlyPlayedPlaylist);
                        SuggestedHeaderView recentlyPlayedHeaderView = new SuggestedHeaderView(recentlyPlayedHeader);
                        recentlyPlayedHeaderView.setClickListener(this);
                        viewModels.add(recentlyPlayedHeaderView);

                        viewModels.addAll(Stream.of(albums)
                                .map(album -> {
                                    AlbumView albumView = new AlbumView(album, ViewType.ALBUM_LIST_SMALL, requestManager);
                                    albumView.setClickListener(this);
                                    return albumView;
                                }).toList());

                        return viewModels;
                    } else {
                        return Collections.emptyList();
                    }
                });
    }

    @SuppressLint("CheckResult")
    Observable<List<ViewModel>> getFavoriteSongViewModels() {

        Observable<List<Song>> favoritesSongs = DataManager.getInstance().getFavoriteSongsRelay()
                .take(20);

        return Observable.combineLatest(
                favoritesSongs,
                Playlist.favoritesPlaylist()
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .toObservable(),
                (songs, playlist) -> {
                    if (!songs.isEmpty()) {
                        List<ViewModel> viewModels = new ArrayList<>();

                        SuggestedHeader favoriteHeader = new SuggestedHeader(getString(R.string.fav_title), getString(R.string.suggested_favorite_subtitle), playlist);
                        SuggestedHeaderView favoriteHeaderView = new SuggestedHeaderView(favoriteHeader);
                        favoriteHeaderView.setClickListener(SuggestedFragment.this);
                        viewModels.add(favoriteHeaderView);

                        viewModels.add(favoriteRecyclerView);

                        SongClickListener songClickListener = new SongClickListener(songs);
                        favoriteRecyclerView.viewModelAdapter.setItems(Stream.of(songs).map(song -> {
                            SuggestedSongView suggestedSongView = new SuggestedSongView(song, requestManager);
                            suggestedSongView.setClickListener(songClickListener);
                            return (ViewModel) suggestedSongView;
                        }).toList());

                        return viewModels;
                    } else {
                        return Collections.emptyList();
                    }
                });
    }

    Observable<List<ViewModel>> getRecentlyAddedViewModels() {
        return Playlist.recentlyAddedPlaylist
                .getSongsObservable()
                .flatMap(songs -> Observable.just(Operators.songsToAlbums(songs)))
                .flatMapSingle(source -> Observable.fromIterable(source)
                        .sorted((a, b) -> ComparisonUtils.compareLong(b.songPlayCount, a.songPlayCount))
                        .take(20)
                        .toList())
                .map(albums -> {
                    if (!albums.isEmpty()) {
                        List<ViewModel> viewModels = new ArrayList<>();

                        SuggestedHeader recentlyAddedHeader = new SuggestedHeader(getString(R.string.recentlyadded), getString(R.string.suggested_recently_added_subtitle), Playlist.recentlyAddedPlaylist);
                        SuggestedHeaderView recentlyAddedHeaderView = new SuggestedHeaderView(recentlyAddedHeader);
                        recentlyAddedHeaderView.setClickListener(this);
                        viewModels.add(recentlyAddedHeaderView);

                        viewModels.addAll(Stream.of(albums).map(album -> {
                            AlbumView albumView = new AlbumView(album, ViewType.ALBUM_CARD, requestManager);
                            albumView.setClickListener(this);
                            return albumView;
                        }).toList());

                        return viewModels;
                    } else {
                        return Collections.emptyList();
                    }
                });
    }

    void refreshAdapterItems() {

        if (setItemsDisposable != null) {
            setItemsDisposable.dispose();
        }

        PermissionUtils.RequestStoragePermissions(() -> {
            if (getActivity() != null && isAdded()) {
                refreshDisposables.add(Observable.combineLatest(
                        getMostPlayedViewModels(),
                        getRecentlyPlayedViewModels(),
                        getFavoriteSongViewModels().switchIfEmpty(Observable.just(Collections.emptyList())),
                        getRecentlyAddedViewModels(),
                        (mostPlayedSongs1, recentlyPlayedAlbums1, favoriteSongs1, recentlyAddedAlbums1) -> {
                            List<ViewModel> items = new ArrayList<>();
                            items.addAll(mostPlayedSongs1);
                            items.addAll(recentlyPlayedAlbums1);
                            items.addAll(favoriteSongs1);
                            items.addAll(recentlyAddedAlbums1);
                            return items;
                        })
                        .debounce(200, TimeUnit.MILLISECONDS)
                        .switchIfEmpty(Observable.just(Collections.emptyList()))
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(adaptableItems -> {
                            if (adaptableItems.isEmpty()) {
                                setItemsDisposable = adapter.setItems(Collections.singletonList((new EmptyView(R.string.empty_suggested))));
                            } else {
                                setItemsDisposable = adapter.setItems(adaptableItems);
                            }
                        }, error -> LogUtils.logException(TAG, "Error setting items", error)));
            }
        });
    }

    @Override
    public void onPause() {

        if (refreshDisposables != null) {
            refreshDisposables.clear();
        }

        super.onPause();
    }

    @Override
    public void onAlbumClick(int position, AlbumView albumView, AlbumView.ViewHolder viewHolder) {
        if (suggestedClickListener != null) {
            suggestedClickListener.onAlbumClicked(albumView.album, viewHolder.imageOne);
        }
    }

    @Override
    public boolean onAlbumLongClick(int position, AlbumView albumView) {
        return false;
    }

    @Override
    public void onAlbumOverflowClicked(View v, Album album) {
        PopupMenu menu = new PopupMenu(getContext(), v);
        MenuUtils.setupAlbumMenu(menu);
        menu.setOnMenuItemClickListener(MenuUtils.getAlbumMenuClickListener(getContext(), album, taggerDialog -> taggerDialog.show(getFragmentManager()), () -> UpgradeDialog.getUpgradeDialog(getActivity()).show()));
        menu.show();
    }

    @Override
    public void onSuggestedHeaderClick(SuggestedHeader suggestedHeader) {
        getNavigationController().pushViewController(PlaylistDetailFragment.newInstance(suggestedHeader.playlist), "PlaylistFragment");
    }

    @Override
    protected String screenName() {
        return TAG;
    }
}

File: app/src/main/java/com/simplecity/amp_library/ui/screens/suggested/SuggestedFragment.kt
package com.simplecity.amp_library.ui.screens.suggested

import android.content.Context
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.bumptech.glide.RequestManager
import com.simplecity.amp_library.R
import com.simplecity.amp_library.R.string
import com.simplecity.amp_library.data.Repository
import com.simplecity.amp_library.model.Album
import com.simplecity.amp_library.model.AlbumArtist
import com.simplecity.amp_library.model.Playlist
import com.simplecity.amp_library.model.Song
import com.simplecity.amp_library.model.SuggestedHeader
import com.simplecity.amp_library.ui.screens.tagger.TaggerDialog
import com.simplecity.amp_library.ui.adapters.ViewType
import com.simplecity.amp_library.ui.common.BaseFragment
import com.simplecity.amp_library.ui.dialog.AlbumBiographyDialog
import com.simplecity.amp_library.ui.dialog.DeleteDialog
import com.simplecity.amp_library.ui.dialog.SongInfoDialog
import com.simplecity.amp_library.ui.modelviews.AlbumView
import com.simplecity.amp_library.ui.modelviews.EmptyView
import com.simplecity.amp_library.ui.modelviews.HorizontalRecyclerView
import com.simplecity.amp_library.ui.modelviews.SuggestedHeaderView
import com.simplecity.amp_library.ui.modelviews.SuggestedSongView
import com.simplecity.amp_library.ui.screens.playlist.detail.PlaylistDetailFragment
import com.simplecity.amp_library.ui.screens.playlist.dialog.CreatePlaylistDialog
import com.simplecity.amp_library.ui.screens.suggested.SuggestedPresenter.SuggestedData
import com.simplecity.amp_library.ui.views.SuggestedDividerDecoration
import com.simplecity.amp_library.utils.ArtworkDialog
import com.simplecity.amp_library.utils.RingtoneManager
import com.simplecity.amp_library.utils.SettingsManager
import com.simplecity.amp_library.utils.ShuttleUtils
import com.simplecity.amp_library.utils.extensions.share
import com.simplecity.amp_library.utils.menu.album.AlbumMenuUtils
import com.simplecity.amp_library.utils.menu.song.SongMenuUtils
import com.simplecity.amp_library.utils.playlists.FavoritesPlaylistManager
import com.simplecity.amp_library.utils.playlists.PlaylistMenuHelper
import com.simplecity.amp_library.utils.sorting.SortManager
import com.simplecity.amp_library.utils.withArgs
import com.simplecityapps.recycler_adapter.adapter.ViewModelAdapter
import com.simplecityapps.recycler_adapter.model.ViewModel
import com.simplecityapps.recycler_adapter.recyclerview.RecyclerListener
import dagger.android.support.AndroidSupportInjection
import io.reactivex.disposables.CompositeDisposable
import java.util.ArrayList
import javax.inject.Inject

class SuggestedFragment :
    BaseFragment(),
    SuggestedHeaderView.ClickListener,
    AlbumView.ClickListener,
    SuggestedContract.View {

    @Inject lateinit var presenter: SuggestedPresenter

    @Inject lateinit var songsRepository: Repository.SongsRepository

    @Inject lateinit var sortManager: SortManager

    @Inject lateinit var settingsManager: SettingsManager

    @Inject lateinit var favoritesPlaylistManager: FavoritesPlaylistManager

    @Inject lateinit var playlistMenuHelper: PlaylistMenuHelper

    @Inject lateinit var requestManager: RequestManager

    private lateinit var adapter: ViewModelAdapter

    private lateinit var favoriteRecyclerView: HorizontalRecyclerView

    private lateinit var mostPlayedRecyclerView: HorizontalRecyclerView

    private val disposables = CompositeDisposable()

    private val refreshDisposables = CompositeDisposable()

    private var suggestedClickListener: SuggestedClickListener? = null

    interface SuggestedClickListener {

        fun onAlbumArtistClicked(albumArtist: AlbumArtist, transitionView: View)

        fun onAlbumClicked(album: Album, transitionView: View)
    }


    // Lifecycle

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)

        if (parentFragment is SuggestedClickListener) {
            suggestedClickListener = parentFragment as SuggestedClickListener?
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        adapter = ViewModelAdapter()
        mostPlayedRecyclerView = HorizontalRecyclerView("SuggestedFragment - mostPlayed")
        favoriteRecyclerView = HorizontalRecyclerView("SuggestedFragment - favorite")
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_suggested, container, false) as RecyclerView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view as RecyclerView

        val spanCount = if (ShuttleUtils.isTablet(context!!)) 12 else 6

        val gridLayoutManager = GridLayoutManager(context, spanCount)
        gridLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                if (!adapter.items.isEmpty() && position >= 0) {
                    val item = adapter.items[position]
                    if (item is HorizontalRecyclerView
                        || item is SuggestedHeaderView
                        || item is AlbumView && item.getViewType() == ViewType.ALBUM_LIST
                        || item is AlbumView && item.getViewType() == ViewType.ALBUM_LIST_SMALL
                        || item is EmptyView
                    ) {
                        return spanCount
                    }
                    if (item is AlbumView && item.getViewType() == ViewType.ALBUM_CARD_LARGE) {
                        return 3
                    }
                }

                return 2
            }
        }

        view.addItemDecoration(SuggestedDividerDecoration(resources))
        view.setRecyclerListener(RecyclerListener())
        view.layoutManager = gridLayoutManager
        view.adapter = adapter

        presenter.bindView(this)
    }

    override fun onResume() {
        super.onResume()

        presenter.loadData()
    }

    override fun onPause() {

        disposables.clear()

        refreshDisposables.clear()

        super.onPause()
    }

    override fun onDestroyView() {
        super.onDestroyView()

        presenter.unbindView(this)
    }

    override fun onDetach() {
        super.onDetach()

        suggestedClickListener = null
    }

    inner class SongClickListener(val songs: List<Song>) : SuggestedSongView.ClickListener {

        override fun onSongClick(song: Song, holder: SuggestedSongView.ViewHolder) {
            mediaManager.playAll(songs, songs.indexOf(song), true) {
                onPlaybackFailed()
            }
        }

        override fun onSongOverflowClicked(v: View, position: Int, song: Song) {
            val popupMenu = PopupMenu(context!!, v)
            SongMenuUtils.setupSongMenu(popupMenu, false, true, playlistMenuHelper)
            popupMenu.setOnMenuItemClickListener(SongMenuUtils.getSongMenuClickListener(song, presenter))
            popupMenu.show()
        }
    }


    // AlbumView.ClickListener implementation

    override fun onAlbumClick(position: Int, albumView: AlbumView, viewHolder: AlbumView.ViewHolder) {
        suggestedClickListener?.onAlbumClicked(albumView.album, viewHolder.imageOne)
    }

    override fun onAlbumLongClick(position: Int, albumView: AlbumView): Boolean {
        return false
    }

    override fun onAlbumOverflowClicked(v: View, album: Album) {
        val menu = PopupMenu(context!!, v)
        AlbumMenuUtils.setupAlbumMenu(menu, playlistMenuHelper, true)
        menu.setOnMenuItemClickListener(AlbumMenuUtils.getAlbumMenuClickListener(album, presenter))
        menu.show()
    }

    override fun onSuggestedHeaderClick(suggestedHeader: SuggestedHeader) {
        navigationController.pushViewController(PlaylistDetailFragment.newInstance(suggestedHeader.playlist), "PlaylistListFragment")
    }


    // SuggestedContract.View implementation

    override fun setData(suggestedData: SuggestedData) {

        val viewModels = ArrayList<ViewModel<*>>()

        if (suggestedData.mostPlayedSongs.isNotEmpty()) {
            val mostPlayedHeader = SuggestedHeader(getString(string.mostplayed), getString(string.suggested_most_played_songs_subtitle), suggestedData.mostPlayedPlaylist)
            val mostPlayedHeaderView = SuggestedHeaderView(mostPlayedHeader)
            mostPlayedHeaderView.setClickListener(this)
            viewModels.add(mostPlayedHeaderView)
            viewModels.add(mostPlayedRecyclerView)

            val songClickListener = SongClickListener(suggestedData.mostPlayedSongs)

            mostPlayedRecyclerView.setItems(suggestedData.mostPlayedSongs
                .map { song ->
                    val suggestedSongView = SuggestedSongView(song, requestManager, settingsManager)
                    suggestedSongView.setClickListener(songClickListener)
                    suggestedSongView
                })
        }

        if (suggestedData.recentlyPlayedAlbums.isNotEmpty()) {
            val recentlyPlayedHeader = SuggestedHeader(getString(string.suggested_recent_title), getString(string.suggested_recent_subtitle), suggestedData.recentlyPlayedPlaylist)
            val recentlyPlayedHeaderView = SuggestedHeaderView(recentlyPlayedHeader)
            recentlyPlayedHeaderView.setClickListener(this)
            viewModels.add(recentlyPlayedHeaderView)

            viewModels.addAll(
                suggestedData.recentlyPlayedAlbums
                    .map { album ->
                        val albumView = AlbumView(album, ViewType.ALBUM_LIST_SMALL, requestManager, sortManager, settingsManager)
                        albumView.setClickListener(this)
                        albumView
                    }.toList()
            )
        }

        if (suggestedData.favoriteSongs.isNotEmpty()) {
            val favoriteSongsHeader = SuggestedHeader(getString(string.fav_title), getString(string.suggested_favorite_subtitle), suggestedData.favoriteSongsPlaylist)
            val favoriteHeaderView = SuggestedHeaderView(favoriteSongsHeader)
            favoriteHeaderView.setClickListener(this)
            viewModels.add(favoriteHeaderView)

            viewModels.add(favoriteRecyclerView)

            val songClickListener = SongClickListener(suggestedData.favoriteSongs)
            analyticsManager.dropBreadcrumb(TAG, "favoriteRecyclerView.setItems()")
            favoriteRecyclerView.setItems(
                suggestedData.favoriteSongs
                    .map { song ->
                        val suggestedSongView = SuggestedSongView(song, requestManager, settingsManager)
                        suggestedSongView.setClickListener(songClickListener)
                        suggestedSongView
                    }.toList()
            )
        }

        if (suggestedData.recentlyAddedAlbums.isNotEmpty()) {
            val recentlyAddedHeader = SuggestedHeader(getString(string.recentlyadded), getString(string.suggested_recently_added_subtitle), suggestedData.recentlyAddedAlbumsPlaylist)
            val recentlyAddedHeaderView = SuggestedHeaderView(recentlyAddedHeader)
            recentlyAddedHeaderView.setClickListener(this)
            viewModels.add(recentlyAddedHeaderView)

            viewModels.addAll(
                suggestedData.recentlyAddedAlbums
                    .map { album ->
                        val albumView = AlbumView(album, ViewType.ALBUM_CARD, requestManager, sortManager, settingsManager)
                        albumView.setClickListener(this)
                        albumView
                    }.toList()
            )
        }

        if (viewModels.isEmpty()) {
            refreshDisposables.add(adapter.setItems(listOf<ViewModel<*>>(EmptyView(R.string.empty_suggested))))
        } else {
            refreshDisposables.add(adapter.setItems(viewModels))
        }
    }


    // SongMenuContract.View Implementation

    override fun presentCreatePlaylistDialog(songs: List<Song>) {
        CreatePlaylistDialog.newInstance(songs).show(childFragmentManager, "CreatePlaylistDialog")
    }

    override fun presentSongInfoDialog(song: Song) {
        SongInfoDialog.newInstance(song).show(childFragmentManager)
    }

    override fun onSongsAddedToPlaylist(playlist: Playlist, numSongs: Int) {
        Toast.makeText(context, context!!.resources.getQuantityString(R.plurals.NNNtrackstoplaylist, numSongs, numSongs), Toast.LENGTH_SHORT).show()
    }

    override fun onSongsAddedToQueue(numSongs: Int) {
        Toast.makeText(context, context!!.resources.getQuantityString(R.plurals.NNNtrackstoqueue, numSongs, numSongs), Toast.LENGTH_SHORT).show()
    }

    override fun presentTagEditorDialog(song: Song) {
        TaggerDialog.newInstance(song).show(childFragmentManager)
    }

    override fun presentDeleteDialog(songs: List<Song>) {
        DeleteDialog.newInstance(DeleteDialog.ListSongsRef { songs }).show(childFragmentManager)
    }

    override fun shareSong(song: Song) {
        song.share(context!!)
    }

    override fun presentRingtonePermissionDialog() {
        RingtoneManager.getDialog(context!!).show()
    }

    override fun showRingtoneSetMessage() {
        Toast.makeText(context, R.string.ringtone_set_new, Toast.LENGTH_SHORT).show()
    }


    // AlbumMenuContract.View Implementation

    override fun onPlaybackFailed() {
        // Todo: Improve error message
        Toast.makeText(context, R.string.emptyplaylist, Toast.LENGTH_SHORT).show()
    }

    override fun presentTagEditorDialog(album: Album) {
        TaggerDialog.newInstance(album).show(childFragmentManager)
    }

    override fun presentDeleteAlbumsDialog(albums: List<Album>) {
        DeleteDialog.newInstance(DeleteDialog.ListAlbumsRef { albums }).show(childFragmentManager)
    }

    override fun presentAlbumInfoDialog(album: Album) {
        AlbumBiographyDialog.newInstance(album).show(childFragmentManager)
    }

    override fun presentArtworkEditorDialog(album: Album) {
        ArtworkDialog.build(context, album).show()
    }


    // BaseFragment implementation

    override fun screenName(): String {
        return TAG
    }


    // Static

    companion object {

        private const val ARG_TITLE = "title"

        private const val TAG = "SuggestedFragment"

        fun newInstance(title: String) = SuggestedFragment().withArgs {
            putString(ARG_TITLE, title)
        }
    }
}
